### PR TITLE
DEPR: Series logical ops with different index coercing to bool

### LINKED ISSUE #52500: BUG: Results of series `bitwise` ufunc operations are being casted to `bool` in `pandas-2.0`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: np.bitwise_xor
Out[3]: <ufunc 'bitwise_xor'>

In [4]: a = pd.Series([1, 2, 3], index=[10, 11, 23], name="a")

In [5]: b = pd.Series([10, 20, 30], index=[11, 10, 23], name="a")

In [9]: np.bitwise_xor(a, b)
Out[9]: 
10    True
11    True
23    True
Name: a, dtype: bool

In [10]: np.bitwise_xor(a.to_frame(), b.to_frame())
Out[10]: 
     a
10  21
11   8
23  29
```

### Issue Description

`bitwise_xor` on `Series` & `DataFrame` are returning different results(i.e., `bool` & `int` respectively).

### Expected Behavior

Consistent output between `Series` & `DataFrame`:
```python
In [9]: np.bitwise_xor(a, b)
Out[9]: 
10    21
11    8
23    29
Name: a, dtype: int64
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : c2a7f1ae753737e589617ebaaff673070036d653
python           : 3.10.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-76-generic
Version          : #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 2.0.0rc1
numpy            : 1.23.5
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.6.1
pip              : 23.0.1
Cython           : 0.29.34
pytest           : 7.2.2
hypothesis       : 6.70.2
sphinx           : 5.3.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.12.0
pandas_datareader: None
bs4              : 4.12.0
bottleneck       : None
brotli           : 
fastparquet      : None
fsspec           : 2023.3.0
gcsfs            : None
matplotlib       : None
numba            : 0.56.4
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 10.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : 2023.3.0
scipy            : 1.10.1
snappy           : 
sqlalchemy       : 1.4.46
tables           : None
tabulate         : 0.9.0
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : None
qtpy             : None
pyqt5            : None

</details>

### LINKED ISSUE #52538: BUG: inconsistent Series/DataFrame behavior in bitwise ops
```python
ser = pd.Series([1, 2], index=['a', 'b'])
ser2 = pd.Series([3, 4], index=['b', 'c'])
df = ser.to_frame()
df2 = ser2.to_frame()

>>> ser & ser2
a    False
b     True
c    False


>>> df & df2
Traceback (most recent call last):
[...]
TypeError: unsupported operand type(s) for &: 'float' and 'bool'
```

Series._logical_method passes `align_asobject=True` to `_align_for_op`, which DataFrame does not.

The net effect of this is in some ways analogous to DataFrame._arith_method doing _arith_method_with_reindex to only operate on the intersection of columns and then reindexing the result.  Maybe we could do something similar on the Series method instead of casting.

Regardless, the behaviors should match.

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #52500 (Replace xxxx with the GitHub issue number)
- [x] closes #52538 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jbrockmendel** [comment]
I for sure prefer this behavior, but its definitely intentional so not sure we can call it a bugfix.

The approach i've been kicking around (with little success) for non-aligned cases is to do a .align(join="inner") before operating, kind of like how we do _arith_method_with_reindex

**Comment by mroeschke** [comment]
Fair point. I went ahead and deprecated this behavior instead

File: pandas/core/series.py [hunk 1]
```diff
@@ -5579,6 +5579,15 @@ def _align_for_op(self, right, align_asobject: bool = False):
...
                 if align_asobject:
+                    if left.dtype != np.bool_ and left.dtype != np.bool_:
+                        warnings.warn(
+                            "Operation between non boolean Series with different "
+                            "indexes will no longer return a boolean result in "
+                            "a future version. Cast both Series to object type "
+                            "to maintain the prior behavior.",
+                            FutureWarning,
+                            stacklevel=find_stack_level(),
+                        )
```
Comment (jbrockmendel) [inline pandas/core/series.py:5590]: FWIW i think we should deprecate this casting in all cases

Comment (mroeschke) [reply pandas/core/series.py:5590]: Agreed, I just wanted to issue to warning though were the result type would change i.e. is a user is doing logical ops on booleans they should already get a boolean result hence no need to warn

**Comment by mroeschke** [comment]
Going to merge, but can address any followups

File: pandas/core/series.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/series.py:5584]: so do we need to do a second deprecation to handle more cases or when this deprecation is enforced will we be able to remove the entire align_asobject block?

**Pull request merged**